env.go

```diff
@@ -19,64 +25,191 @@ type EnvVar struct {
+}
+
+// GetEnv returns the value of the environment variable given with key.
+// The variable is assumed to have been registered either with E or
+// imported from existing environment
+func GetEnv(key string) (value string, exists bool) {
+	return env.GetEnv(key)
+}
+
+// Env returns the complete environment
+func Env() map[string]EnvVar {
+	return env.Env()
+}
+
+// NewEnviron creates a new configuration environment
+func NewEnviron(importer EnvImporterFunc) *Environ {
+	env = newEnviron(importer)
+	return env
+}
+
+// E defines a new environment variable specified with e.
+// Returns the current value of the variable with precedence
+// given to previously imported environment variables.
+// If the variable was not previously imported and no value
+// has been specified, the default is returned
+func (r *Environ) E(e EnvVar) string {
 	if e.Key == "" {
-		panic("Key shouldn't be empty")
+		panic("key shouldn't be empty")
 	}
-
 	if e.Secret && len(e.Default) > 0 {
-		panic("Secrets shouldn't be embedded with defaults")
+		panic("secrets shouldn't be embedded with defaults")
 	}
 
-	if EnvVars == nil {
-		EnvVars = make(map[string]EnvVar)
-		ImportEnvVars = make(map[string]string)
+	r.importOnce()
+	if v, ok := r.imported[e.Key]; ok {
+		e.Value = v
+	} else {
+		e.Value = os.Getenv(e.Key)
 	}
-	e.Value = os.Getenv(e.Key)
-
-	EnvVars[e.Key] = e
+	r.env[e.Key] = e
 
-	return GetEnv(e.Key)
+	return r.MustGetEnv(e.Key)
 }
 
-func GetEnv(key string) string {
-	importMakeEnv()
+// MustGetEnv returns the value of the environment variable given with key.
+// The variable is assumed to have been registered either with E or
+// imported from existing environment - otherwise the function will panic.
+// For non-panicking version use GetEnv
+func (r *Environ) MustGetEnv(key string) (value string) {
+	if v, ok := r.GetEnv(key); ok {
+		return v
+	}
+	panic(fmt.Sprintf("Requested environment variable %q hasn't been registered", key))
+}
 
-	if EnvVars == nil {
-		EnvVars = make(map[string]EnvVar)
-		ImportEnvVars = make(map[string]string)
+// GetEnv returns the value of the environment variable given with key.
+// The variable is assumed to have been registered either with E or
+// imported from existing environment
+func (r *Environ) GetEnv(key string) (value string, exists bool) {
+	r.importOnce()
+	var v EnvVar
+	if v, exists = r.env[key]; !exists {
+		return "", false
 	}
+	if v.Value != "" {
+		return v.Value, true
+	}
+	return v.Default, true
+}
 
-	if v, ok := EnvVars[key]; ok {
-		if v.Value != "" {
-			return v.Value
+// Env returns the complete environment
+func (r *Environ) Env() map[string]EnvVar {
+	m := make(map[string]EnvVar, len(r.env))
+	for key, value := range r.env {
+		def := value.Default
+		if def == "" {
+			def = r.imported[key]
 		}
+		value.Default = def
+		m[key] = value
+	}
+	return m
+}
 
-		if v, ok := ImportEnvVars[key]; ok {
-			return v
-		}
+// Environ represents the environment with configuration
```

Comment:
Had to introduce an additional type for handling configuration from environment and allow the option to read overrides from an arbitrary source (e.g. in my case a custom makefile).